### PR TITLE
Issue #1693336 by chx, David_Rothstein, catch | iamEAP: Block rehashing happens on every cron run.

### PR DESCRIPTION
**Pull request by hellomrcat**
PR for https://github.com/backdrop/backdrop-issues/issues/222

### CODE REVIEW

File: core/modules/block/block.module [hunk 1]
```diff
@@ -459,8 +462,15 @@ function _block_rehash($theme = NULL) {
...
-      // Add to the list of blocks we return.
+      // If the block is new or differs from the original database block, save
+      // it. To determine whether there was a change it is enough to examine
+      // the values for the keys in the original database record as that
+      // contained every database field.
+      if (!$primary_keys || array_diff_assoc($original_database_blocks[$module][$delta], $block)) {
+        backdrop_write_record('block', $block, $primary_keys);
+        // Make it possible to test this.
+        $block['saved'] = TRUE;
+      }      // Add to the list of blocks we return.
```
Comment (quicksketch) [inline core/modules/block/block.module]: Let's put this code comment on the next line where it was before.

File: core/modules/block/block.module [hunk 2]
```diff
@@ -404,23 +404,26 @@ function _block_rehash($theme = NULL) {
...
     ->condition('theme', $theme)
     ->execute();
+  $original_database_blocks = array();
   foreach ($database_blocks as $block) {
-    // Preserve info which is not in the database.
-    $block->info = $current_blocks[$block->module][$block->delta]['info'];
-    // The cache mode can only by set from hook_block_info(), so that has
+    $module = $block['module'];
+    $delta = $block['delta'];
+    $original_database_blocks[$module][$delta] = $block;    // The cache mode can only by set from hook_block_info(), so that has
```
Comment (quicksketch) [inline core/modules/block/block.module]: Code comment should be on the next line here.

File: core/modules/block/tests/block_test.module [hunk 3]
```diff
@@ -34,3 +34,12 @@ function block_test_block_info() {
 function block_test_block_view($delta = 0) {
   return array('content' => state_get('block_test_content', ''));
 }
+
+/**
+ * Implements hook_block_info_alter().
+ */
+function block_test_block_info_alter(&$blocks) {
+  if (variable_get('block_test_info_alter')) {
```
Comment (quicksketch) [inline core/modules/block/tests/block_test.module]: Let's not introduce any new variable_get() calls. Can this be converted to use state_get()?

File: core/modules/block/block.test [hunk 4]
```diff
@@ -827,3 +827,81 @@ class BlockInvalidRegionTestCase extends BackdropWebTestCase {
...
+ * Tests that block rehashing works correctly.
+ */
+class BlockHashTestCase extends backdropWebTestCase {
+  public static function getInfo() {
+    return array(
+      'name' => 'Block rehash',
+      'description' => 'Checks _block_rehash() functionality.',
+      'group' => 'Block',
+    );
+  }
```
Comment (quicksketch) [inline core/modules/block/block.test]: This getInfo() method needs to be moved to block.tests.info.

File: core/modules/block/block.test [hunk 5]
```diff
@@ -827,3 +827,81 @@ class BlockInvalidRegionTestCase extends BackdropWebTestCase {
...
+    // Save the new blocks, check that the new blocks exist by checking weight.
+    _block_rehash();
+    $this->assertWeight(0);
+    // Now hook_block_info_alter() exists but no blocks are saved on a second
+    // rehash.
+    $this->doRehash();
+    $this->assertWeight(0);
+    // Now hook_block_info_alter() exists and is changing one block which
+    // should be saved.
+    $GLOBALS['conf']['block_test_info_alter'] = 1;
```
Comment (quicksketch) [inline core/modules/block/block.test]: Use state_set() instead of $GLOBALS['conf'] here (which is a hack to inject into variable_get()).

**Comment by hellomrcat** [comment]
@quicksketch thanks for the help!

File: core/modules/block/block.test [hunk 6]
```diff
@@ -827,3 +827,73 @@ class BlockInvalidRegionTestCase extends BackdropWebTestCase {
     $this->assertNoRaw($warning_message, 'Disabled block in the invalid region will not trigger the warning.');
   }
 }
+
+/**
+ * Tests that block rehashing works correctly.
+ */
+class BlockHashTestCase extends backdropWebTestCase {
```
Comment (quicksketch) [inline core/modules/block/block.test]: Although minor, `backdropWebTestCase` needs to be `backdropWebTestCase` to match the actual class name, but PHP isn't case sensitive so it's not throwing any errors.

**Comment by jenlampton** [comment]
Looks like this one is good to go now.

**Comment by quicksketch** [comment]
Super!

**Pull request merged**